File: 1-working-with-list/p05_rev.ml

let rec rev = function
    [] -> []
  | hd :: tl -> (rev tl) @ [hd];;

rev ["a" ; "b" ; "c"];;
rev ["a"];;
rev [];;

let rev l =
  let rec call acc = function
    | [] -> acc
    | hd :: tl -> call (hd::acc) tl
  in
  call [] l;;

rev ["a" ; "b" ; "c"];;
rev ["a"];;
rev [];;

File: 1-working-with-list/p05_rev.scala

object Main {

  def main(args: Array[String]) {
    println(rev(List("a", "b", "c")))
  }

  def rev[T](l: List[T]): List[T] = {
    l match {
      case Nil => Nil
      case hd :: rest => rev(rest) :+ hd
    }
  }
}
